### ISSUE TITLE
`azd down` with a VNET/NSG, sometimes fails because NSG can't be deleted first.

### ISSUE THREAD

**Issue by adrianhall** [Azure/azure-dev#2497]
- [X] Make sure you've installed the latest version using [instructions in the wiki](../wiki/install)

**Output from `azd version`**
Run `azd version` and copy and paste the output here:
`azd version 1.0.2 (commit 145e046b1ea9394bd4e1b1d539eb32e860d692fb)`

**Describe the bug**
I have a template that creates a VNET and a set of NSGs, using bicep as the infra language.  When using "azd down", the order of operations is important.  You cannot delete the NSGs prior to the VNET because you get an error

```
ERROR: deleting infrastructure: error deleting Azure resources: deleting resource groups: deleting resource group: GET https://management.azure.com/subscriptions/XXX/operationresults/eyJqb2JJZCI6IlJFU09VUkNFR1JPVVBERUxFVElPTkpPQi1SRzoyREFIOjJEMDcxMDoyREhVQjoyRERFVjoyREFVU1RSQUxJQUVBU1Q6MkRTUE9LRS1BVVNUUkFMSUFFQVNUIiwiam9iTG9jYXRpb24iOiJhdXN0cmFsaWFlYXN0In0
--------------------------------------------------------------------------------
RESPONSE 409: 409 Conflict
ERROR CODE: ResourceGroupDeletionBlocked
--------------------------------------------------------------------------------
```

**To Reproduce**
Clone https://github.com/adrianhall/field-engineer
Run the following:

```shell
azd env new
azd env set NETWORK_ISOLATION true
azd env set DEPLOY_HUB_NETWORK true
azd provision
```

Once deployed, run `azd down` - the error is intermittent since it's due to an ordering issue.

**Expected behavior**
I expect `azd down` to delete the VNET first and the NSGs second.

**Comment by adrianhall** [comment]
To follow-on, there is actually an ordering when using split resource groups.

I believe the issue is that my App Service (in later resource group) is VNET integrated, which causes the VNET to not be able to be deleted, which causes the NSG to not be deleted.

**Comment by ellismg** [comment]
> To follow-on, there is actually an ordering when using split resource groups.

Very likely - the way `azd down` works today is that it discovers all the resource groups from the most recent deployment and then deletes each one individually (and I don't even think the order in which these deletes are scheduled is stable, which is quite unfortunate!).

> I believe the issue is that my App Service (in later resource group) is VNET integrated, which causes the VNET to not be able to be deleted, which causes the NSG to not be deleted

I'm wondering if it would be possible for us to discover these cross resource group dependencies and schedule our deletions in such a way that we wouldn't run into this, assuming there are no cycles (which it seems like there could be in the limit).

Or perhaps there's some way to encode a dependency delete order in the template, or something similar.

@adrianhall - Looking briefly at your template it looks like there is some general teardown order that you as the template author could propose.  I'm guessing that when it comes to deleting these resource groups, you would want to tear down the spoke resource group before the hub resource group and ensure both of them are gone before you delete the main resource group (the one that is not conditional?)

**Comment by adrianhall** [comment]
Yeah - I am producing a "cleanup.ps1", which encodes everything. The things that need special treatment:

1. diagnostic settings for all resources
2. budgets associated with a resource group
3. private endpoints
4. app services (and other services that are vnet integrated or injected)
5. peerings between virtual networks
6. things that need to be purged (api management, key vault, unless included in #4)

At this point, I believe the resource groups can all be deleted in parallel.

**Comment by torumakabe** [comment]
We're experiencing the same dependency ordering issue with Azure Monitor Workspace (AMW) deployments. When running azd down, the command attempts to delete AMW's managed resource groups before deleting the AMW resource itself, resulting in DenyAssignmentAuthorizationFailed errors.

Would appreciate prioritizing this dependency resolution enhancement, as it affects the core reliability of the azd workflow.

**Comment by spboyer** [comment]
Closing as stale — this issue has had no activity for 180+ days. If this is still relevant, please reopen with reproduction steps on the latest version of azd.
